nodejs/awsx/apigateway/cognitoAuthorizer.ts

```diff
@@ -0,0 +1,133 @@
+     * "resource-server/scope". e.g. ["com.hamuta.movies/drama.view",
+     * "http://my.resource.com/file.read"] For more information on resource servers and custom
+     * scopes visit the AWS documentation -
+     * https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-define-resource-servers.html
+     */
+    methodsToAuthorize?: string[];
+}
+
+/**
+ * The set of arguments for constructing a CognitoAuthorizer resource.
+ */
+export interface CognitoAuthorizerArgs {
+
+    /**
+     * The name for the Authorizer to be referenced as. This must be unique for each unique
+     * authorizer within the API. If no name if specified, a name will be generated for you.
+     */
+    authorizerName?: string;
+
+    /**
+     * The request header for the authorization token. If not set, this defaults to
+     * "Authorization".
+     */
+    header?: string;
+
+    /**
+     * The ARNs of the Cognito User Pools to use.
+     */
+    providerARNs: Array<pulumi.Input<string> | aws.cognito.UserPool>;
+
+    /**
+     * A regular expression for validating the token as the incoming identity. It only invokes the
+     * authorizer if there is a match, else it will return a 401. Example: "^x-[a-z]+"
+     */
+    identityValidationExpression?: string;
+
+    /**
+     * The number of seconds during which the resulting IAM policy is cached. Default is 300s. You
+     * can set this value to 0 to disable caching. Max value is 3600s. Note - if you are sharing an
+     * authorizer across more than one route you will want to disable the cache or else it will
+     * cause problems for you.
+     */
+    authorizerResultTtlInSeconds?: number;
+
+    /**
+     * For method authorization, you can define resource servers and custom scopes by specifying the
+     * "resource-server/scope". e.g. ["com.hamuta.movies/drama.view",
+     * "http://my.resource.com/file.read"] For more information on resource servers and custom
+     * scopes visit the AWS documentation -
+     * https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-define-resource-servers.html
+     */
+    methodsToAuthorize?: string[];
+}
+
+/**
+ * getCognitoAuthorizer is a helper function to generate a CognitoAuthorizer.
+ * @param name - the name for the authorizer. This must be unique for each unique authorizer in the API.
+ * @param args - configuration information for the Cognito Authorizer.
+ */
+export function getCognitoAuthorizer(args: CognitoAuthorizerArgs): CognitoAuthorizer {
+    const parameterName = args.header || "Authorization";
+
+    return {
+        authorizerName: args.authorizerName,
+        parameterName: parameterName,
+        providerARNs: args.providerARNs,
+        identitySource: ["method.request.header." + parameterName],
+        identityValidationExpression: args.identityValidationExpression,
+        authorizerResultTtlInSeconds: args.authorizerResultTtlInSeconds,
+        methodsToAuthorize: args.methodsToAuthorize,
```

Comment:
technically can be simplified to:

```ts
return {
    ...args,
    identitySource: ["method.request.header." + parameterName],
};
```

Reply:
parameterName: parameterName cause a problem here... but this is a cool tip!

Reply:
you're right, you'd need to write:

```ts
return {
    ...args,
    parameterName,
    identitySource: ["method.request.header." + parameterName],
};
```